### DISCUSSION TITLE
Modify websocket messages

### DISCUSSION THREAD

**Discussion by LazyCat2** [abhinavsingh/proxy.py#913] [Q&A] [answered]
Hello, 
I want to check and modify websocket messages (not http requests). How can i do that?

**Comment by abhinavsingh** [comment, marked as answer]
Does something like this work for you, can you double check?  You should see `modified websocket response` in websocket responses on the client side.

```python
class ModifyWebsocketResponsePlugin(HttpProxyBasePlugin):
    """Inspect/Modify/Send custom websocket responses."""

    def handle_upstream_chunk(self, chunk: memoryview) -> memoryview:
        remaining = chunk.tobytes()
        while len(remaining) > 0:
            response = WebsocketFrame()
            remaining = response.parse(remaining)
            self.client.queue(memoryview(
                WebsocketFrame.text(b"modified websocket response")
            ))
        return memoryview(b'')
```

If it works, we should check-in this plugin as a starting point within the repo, probably with some documentation.

Let me know.

**Reply by abhinavsingh** [reply]
@Woolbex Sent out a PR https://github.com/abhinavsingh/proxy.py/pull/914 , but let's confirm if this works fine.  I didn't get a chance to try it myself.  Thank you!!!

**Reply by LazyCat2** [reply]
where to import `WebsocketFrame` from?

**Reply by abhinavsingh** [reply]
You can copy the plugin file from the PR instead. See https://github.com/abhinavsingh/proxy.py/pull/914/files#diff-b120ac8759b0ec4c3fb55c25bd010bfac4664c6bd03327dbdf2f33aac7b8e029

`WebsocketFrame` is within `from proxy.http.websocket import WebsocketFrame`

**Reply by LazyCat2** [reply]
i'm getting this error when websocket message sent:
```
Exception in thread Thread-364 (run):
Traceback (most recent call last):
  File "C:\Python310\lib\threading.py", line 1009, in _bootstrap_inner
    self.run()
  File "C:\Python310\lib\threading.py", line 946, in run
    self._target(*self._args, **self._kwargs)
  File "C:\Python310\lib\site-packages\proxy\http\handler.py", line 365, in run
    self.shutdown()
  File "C:\Python310\lib\site-packages\proxy\http\handler.py", line 167, in shutdown
    self.flush()
  File "C:\Python310\lib\site-packages\proxy\http\handler.py", line 210, in flush
    self.selector.register(
  File "C:\Python310\lib\selectors.py", line 300, in register
    key = super().register(fileobj, events, data)
  File "C:\Python310\lib\selectors.py", line 242, in register
    raise KeyError("{!r} (FD {}) is already registered"
KeyError: "<socket.socket fd=440, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=0, laddr=('192.168.8.102', 1234), raddr=('192.168.8.102', 62342)> (FD 440) is already registered"
```

**Reply by abhinavsingh** [reply]
Hey looks like I missed your message totally.  Kindly try latest `RC` release, it fixed a few things related to Windows.  Kindly let me know.

**Reply by mattshep** [reply]
Is it possible to read/modify the messages sent by the client to the upstream?

**Reply by abhinavsingh** [reply]
@mattshep on_client_data callback should give you that opportunity.
